File: ts-others/lib/PolicyChecker.ts
import { Annotations, IAspect, Stack } from "aws-cdk-lib";
import { CfnPolicy } from "aws-cdk-lib/aws-iam";
import { IConstruct } from "constructs";



export class PolicyChecker implements IAspect{

    visit(node: IConstruct): void {
        if (node instanceof CfnPolicy) {
            const resolvedDoc = Stack.of(node).resolve(node.policyDocument)
            const resolvedDocJson = JSON.stringify(resolvedDoc);

            const forbiddenAction = 'GetBucket'

            if (resolvedDocJson.search(forbiddenAction)) {
                Annotations.of(node).addWarningV2('warining',
                'Forbidden action: ' + forbiddenAction
                )
            }
        }
    }
    
}

File: py-others/py_others/policy_checker.py
import jsii
from aws_cdk import IAspect, aws_iam, Stack, Annotations
import json

@jsii.implements(IAspect)
class PolicyChecker:

    def visit(self, node):
        # print(f'Visiting {node.__class__.__name__}')

        if isinstance(node, aws_iam.CfnPolicy):
            resolvedDoc = Stack.of(node).resolve(node.policy_document)
            resolvedDocJson = json.dumps(resolvedDoc)

            forbidden_action = 'GetBucket'

            if forbidden_action in resolvedDocJson:
                Annotations.of(node).add_warning_v2('warning', 
                    'Forbidden action ' + forbidden_action
                    )


